backend/src/main/java/fk/prof/backend/model/request/RecordedProfileParser.java

```diff
@@ -0,0 +1,133 @@
+package fk.prof.backend.model.request;
+
+import com.google.protobuf.CodedInputStream;
+import fk.prof.backend.aggregator.AggregationWindow;
+import fk.prof.backend.exception.AggregationFailure;
+import fk.prof.backend.exception.HttpFailure;
+import fk.prof.backend.service.IProfileWorkService;
+import io.vertx.core.buffer.Buffer;
+import recording.Recorder;
+
+import java.io.IOException;
+import java.time.Clock;
+import java.time.LocalDateTime;
+
+public class RecordedProfileParser {
+
+  private IProfileWorkService profileWorkService;
+  private RecordedProfileHeader header = null;
+  private AggregationWindow aggregationWindow = null;
+  private long workId = 0;
+  private LocalDateTime startedAt = null;
+
+  private RecordedProfileHeaderParser headerParser;
+  private WseParser wseParser;
+  private RecordedProfileIndexes indexes = new RecordedProfileIndexes();
+
+  private boolean intermediateWseEntry = false;
+
+  public RecordedProfileParser(IProfileWorkService profileWorkService, int maxAllowedBytesForRecordingHeader, int maxAllowedBytesForWse) {
+    this.profileWorkService = profileWorkService;
+    this.headerParser = new RecordedProfileHeaderParser(maxAllowedBytesForRecordingHeader);
+    this.wseParser = new WseParser(maxAllowedBytesForWse);
+  }
+
+  /**
+   * Returns true if header has been successfully parsed and no wse entry log is parsed partially
+   *
+   * @return parsed a valid recorded profile object or not
+   */
+  public boolean isParsed() {
+    return headerParser.isParsed() && !intermediateWseEntry;
+  }
+
+  /**
+   * Returns {@link RecordedProfileHeader} if {@link #isParsed()} is true, null otherwise
+   *
+   * @return
+   */
+  public RecordedProfileHeader getProfileHeader() {
+    return headerParser.get();
+  }
+
+  /**
+   * Reads buffer and updates internal state with parsed fields. Returns the starting unread position in outputstream
+   *
+   * @param codedInputStream
+   * @return starting unread position in buffer
+   */
+  public int parse(CodedInputStream codedInputStream, Buffer underlyingBuffer, int currentPos) throws HttpFailure, AggregationFailure {
+    if (startedAt == null) {
+      startedAt = LocalDateTime.now(Clock.systemUTC());
+    }
+
+    try {
+      if (!headerParser.isParsed()) {
+        currentPos = headerParser.parse(codedInputStream, underlyingBuffer, currentPos);
+
+        if (headerParser.isParsed()) {
+          header = headerParser.get();
+
+          workId = header.getRecordingHeader().getWorkAssignment().getWorkId();
+          aggregationWindow = profileWorkService.getAssociatedAggregationWindow(workId);
+          if (aggregationWindow == null) {
+            throw new HttpFailure(String.format("workId=%d not found, cannot continue receiving associated profile",
```

Comment:
RecordedProfileParser shouldn't be throwing HttpFailure. Because, this layer shouldn't be aware of the fact that it is working inside an HTTP container.

It should throw domain oriented exceptions, we can handle them at vertx handler level to translate to HTTP exceptions if need be.

It should probably throw an IllegalStateException here?